flow/alerting/classifier.go

```diff
@@ -0,0 +1,134 @@
+package alerting
+
+import (
+	"context"
+	"errors"
+	"io"
+	"strings"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+	"github.com/jackc/pgx/v5/pgconn"
+)
+
+type ErrorAction string
+
+const (
+	NotifyUser      ErrorAction = "notify_user"
+	Ignore          ErrorAction = "ignore"
+	NotifyTelemetry ErrorAction = "notify_telemetry"
+)
+
+func (e ErrorAction) String() string {
+	return string(e)
+}
+
+type ErrorClass struct {
+	Class  string
+	action ErrorAction
+}
+
+var (
+	ErrorNotifyOOM = ErrorClass{
+		// ClickHouse Code 241
+		Class: "NOTIFY_OOM", action: NotifyUser,
+	}
+	ErrorNotifyMVOrView = ErrorClass{
+		// ClickHouse Code 349 / Code 48 with "while pushing to view", TODO(242,236)
+		Class: "NOTIFY_MV_OR_VIEW", action: NotifyUser,
+	}
+	ErrorNotifyConnectivity = ErrorClass{
+		// ClickHouse Code 81 or Postgres Code 28P01
+		Class: "NOTIFY_CONNECTIVITY", action: NotifyUser,
+	}
+	ErrorNotifySlotInvalid = ErrorClass{
+		// Postgres Code 55000 with "cannot read from logical replication slot"
+		Class: "NOTIFY_SLOT_INVALID", action: NotifyUser,
+	}
+	ErrorNotifyTerminate = ErrorClass{
+		// Postgres Code 57P01
+		Class: "NOTIFY_TERMINATE", action: NotifyUser,
+	}
+	ErrorNotifyConnectTimeout = ErrorClass{
+		// TODO
+		Class: "NOTIFY_CONNECT_TIMEOUT", action: NotifyUser,
+	}
+	ErrorEventInternal = ErrorClass{
+		// Level <= Info
+		Class: "EVENT_INTERNAL", action: NotifyTelemetry,
+	}
+	ErrorIgnoreEOF = ErrorClass{
+		// io.EOF || io.ErrUnexpectedEOF
+		Class: "IGNORE_EOF", action: Ignore,
+	}
+	ErrorIgnoreContextCancelled = ErrorClass{
+		// context.Canceled
+		Class: "IGNORE_CONTEXT_CANCELLED", action: Ignore,
+	}
+	ErrorInternalClickHouse = ErrorClass{
+		// Code 999 or 341
+		Class: "INTERNAL_CLICKHOUSE", action: NotifyTelemetry,
+	}
+	ErrorOther = ErrorClass{
+		// These are internal and should not be exposed
+		Class: "OTHER", action: NotifyTelemetry,
+	}
+)
+
+func (e ErrorClass) String() string {
+	return e.Class
+}
+func (e ErrorClass) ErrorAction() ErrorAction {
+	if e.action != "" {
+		return e.action
+	}
+	return NotifyTelemetry
+}
+
+func GetErrorClass(ctx context.Context, err error) ErrorClass {
```

Comment:
```suggestion
func GetErrorClass(ctx context.Context, err error) (ErrorClass, []string) {
```
this might as well compute tags, lots of overlap

Reply:
> Appending logic currently will cause duplicate tagging

Duplicate tags are fine, will do the same in phases, will have a different prefix for classification and action

> Logic a bit annoying with early returns here

Did it to prevent going through all if cases, can remove the return from switch and move after the switches, IIRC the switches use binary search